gtk4/src/subclass/widget.rs

```diff
@@ -876,6 +880,82 @@ pub unsafe trait WidgetClassSubclassExt: ClassStruct {
         }
     }
 
+    #[doc(alias = "gtk_widget_class_install_action")]
+    fn install_action<
+        F: Fn(&Widget, &str, Option<&Variant>) + 'static,
+        S: WidgetImpl + ObjectSubclass,
+    >(
+        &mut self,
+        action_name: &str,
+        parameter_type: Option<&str>,
+        activate: F,
+    ) {
+        // We store the activate callbacks in a HashMap<action_name, activate>
+        // so that we can retrieve f later on the activate_trampoline call
+        let mut data = <Self::Type as ObjectSubclass>::type_data();
+        let data = data.as_mut();
+
+        #[derive(Debug)]
+        struct Actions(pub(crate) HashMap<String, glib::ffi::gpointer>);
+        unsafe impl Sync for Actions {}
+        unsafe impl Send for Actions {}
+
+        let f: Box_<F> = Box_::new(activate);
+
+        let actions = data
+            .get_class_data_mut::<Actions>(Widget::static_type())
+            .unwrap();
+        actions.0.insert(action_name.to_string(), Box::into_raw(f));
+
+        unsafe extern "C" fn activate_trampoline<F, S>(
+            this: S::Type,
```

Comment:
This currently fails because `the associated type 'Type' is not found` for S, which is defined as `ObjectSubclass`

Reply:
You need to add a bound for `ObjectSubclass` to the function for that to work. Type parameters on nested functions need to repeat all bounds they need, you don't get the type parameters of the surrounding scope for free :)

Nested functions are just like normal functions elsewhere, the only difference is the scoping of the function itself.